### PR TITLE
Inference to use provider resource id to register and validate

### PR DESCRIPTION
**Pull request by dineshyv**
This PR changes the way model id gets translated to the final model name that gets passed through the provider. 
Major changes include:
1) Providers are responsible for registering an object and as part of the registration returning the object with the correct provider specific name of the model provider_resource_id
2) To help with the common look ups different names a new ModelLookup class is created.

Tested all inference providers including together, fireworks, vllm, ollama, meta reference and bedrock

### CODE REVIEW

**Review by raghotham — COMMENTED** [review]
can you make sure to change the zero-to-hero notebooks as well?

File: llama_stack/providers/remote/inference/vllm/vllm.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ async def initialize(self) -> None:
         self.client = OpenAI(base_url=self.config.url, api_key=self.config.api_token)
 
     async def register_model(self, model: Model) -> None:
-        raise ValueError("Model registration is not supported for vLLM models")
+        pass
```
[the remark this answers is not in the document]
Comment (ashwinb) [reply llama_stack/providers/remote/inference/vllm/vllm.py:48]: I am changing this independently to actually query against what vllm is listing

[the remark this answers is not in the document]
Comment (ashwinb) [reply llama_stack/providers/remote/inference/vllm/vllm.py:48]: @yanxi0830 https://github.com/meta-llama/llama-stack/commit/afe4a53ae85466ab8cae220ca11600aa8b1d1a1e

File: llama_stack/providers/remote/inference/vllm/vllm.py [hunk 2]
```diff
@@ -88,8 +84,10 @@ async def chat_completion(
         stream: Optional[bool] = False,
         logprobs: Optional[LogProbConfig] = None,
     ) -> AsyncGenerator:
+        model = await self.model_store.get_model(model_id)
+        print(f"model={model}")
```
Comment (yanxi0830) [inline llama_stack/providers/remote/inference/vllm/vllm.py:88]: remove?

File: llama_stack/providers/tests/inference/test_text_inference.py [hunk 3]
```diff
@@ -171,7 +171,7 @@ async def test_chat_completion_non_streaming(
     ):
         inference_impl, _ = inference_stack
         response = await inference_impl.chat_completion(
-            model=inference_model,
+            model_id=inference_model,
```
Comment (yanxi0830) [inline llama_stack/providers/tests/inference/test_text_inference.py:174]: https://github.com/meta-llama/llama-stack/blob/e51107e019771909b8154b2f024448bb41c45870/llama_stack/providers/inline/eval/meta_reference/eval.py#L152-L154

/eval also calls chat_completion, which also need to be changed?

File: llama_stack/apis/inference/inference.py [hunk 4]
```diff
@@ -237,7 +237,7 @@ async def completion(
     @webmethod(route="/inference/chat_completion")
     async def chat_completion(
         self,
-        model: str,
+        model_id: str,
```
Comment (yanxi0830) [inline llama_stack/apis/inference/inference.py:240]: `llama-stack-apps` examples also needs to be updated: https://github.com/meta-llama/llama-stack-apps/blob/0dc9c42fb42bf21d35e6d231afc4e0360a9eac61/examples/inference/client.py#L46-L49

Comment (yanxi0830) [reply llama_stack/apis/inference/inference.py:240]: https://llama-stack.readthedocs.io/en/latest/getting_started/index.html#chat-completion-test --> also these docs

File: llama_stack/providers/remote/inference/vllm/vllm.py (hunk 2 above)
Comment (dineshyv) [reply llama_stack/providers/remote/inference/vllm/vllm.py:88]: done.

**Review by raghotham — APPROVED** [review]

File: llama_stack/providers/tests/inference/test_text_inference.py (hunk 3 above)
Comment (dineshyv) [reply llama_stack/providers/tests/inference/test_text_inference.py:174]: done.

File: llama_stack/providers/utils/inference/model_registry.py [hunk 5]
```diff
@@ -4,32 +4,75 @@
...
+        llama_model=model_descriptor,
+    )
 
-        return self.stack_to_provider_models_map[identifier]
 
-    async def register_model(self, model: Model) -> None:
-        if model.identifier not in self.stack_to_provider_models_map:
-            raise ValueError(
-                f"Unsupported model {model.identifier}. Supported models: {self.stack_to_provider_models_map.keys()}"
+class ModelLookup:
```
Comment (ashwinb) [inline llama_stack/providers/utils/inference/model_registry.py:36]: why create another class? can this go inside the registry helper directly?

**Comment by ashwinb** [comment]
very useful change!

File: llama_stack/apis/inference/inference.py (hunk 4 above)
Comment (dineshyv) [reply llama_stack/apis/inference/inference.py:240]: done.

Comment (yanxi0830) [reply llama_stack/apis/inference/inference.py:240]: llama-stack-apps and llama-stack-client-python also needs to be updated to reflect the `model` -> `model_id` change.  

https://github.com/meta-llama/llama-stack-apps/blob/0dc9c42fb42bf21d35e6d231afc4e0360a9eac61/examples/inference/client.py#L46-L49

**Pull request merged**